activate.go

```diff
@@ -0,0 +1,665 @@
+ *
+ */
+
+package secboot
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"sort"
+
+	"golang.org/x/sys/unix"
+)
+
+// ErrCannotActivate is returned from ActivateContext.ActivateContainer
+// if a storage container cannot be activated because there are no valid
+// keyslots and / or there are no more passphrase, PIN or recovery key
+// attempts remaining.
+var ErrCannotActivate = errors.New("cannot activate: no valid keyslots and / or no more passphrase, PIN or recovery key tries remaining")
+
+type errorKeyslot struct {
+	slotType KeyslotType
+	name     string
+}
+
+func (i *errorKeyslot) Type() KeyslotType {
+	return i.slotType
+}
+
+func (i *errorKeyslot) Name() string {
+	return i.name
+}
+
+func (i *errorKeyslot) Priority() int {
+	return 0
+}
+
+func (i *errorKeyslot) Data() KeyDataReader {
+	return nil
+}
+
+// keyslotAttemptRecord binds together information about a keyslot.
+type keyslotAttemptRecord struct {
+	slot Keyslot  // The backend supplied Keyslot.
+	data *KeyData // A cache of the decoded KeyData for platform keys.
+	err  error    // The first error that occurred with this keyslot.
+}
+
+func (r *keyslotAttemptRecord) usable() bool {
+	if r.err == nil {
+		return true
+	}
+	return errors.Is(r.err, ErrInvalidPassphrase)
+}
+
+type keyslotAttemptRecordSlice []*keyslotAttemptRecord
+
+func (s keyslotAttemptRecordSlice) Len() int {
+	return len(s)
+}
+
+func (s keyslotAttemptRecordSlice) Less(i, j int) bool {
+	switch {
+	case s[i].slot.Priority() != s[j].slot.Priority():
+		// Order higher priority keyslots first.
+		return s[i].slot.Priority() > s[j].slot.Priority()
+	default:
+		// Order keyslots with the same priority in
+		// name order.
+		return s[i].slot.Name() < s[j].slot.Name()
+	}
+}
+
+func (s keyslotAttemptRecordSlice) Swap(i, j int) {
+	tmp := s[j]
+	s[j] = s[i]
+	s[i] = tmp
+}
+
+// activateOneContainerStateMachineTask describes a state of
+// the state machine, including a readable name and a callback function.
+type activateOneContainerStateMachineTask struct {
+	name string
+	fn   func(context.Context) error
+}
+
+// activateOneContainerStateMachine is a state machine for activating
+// a single StorageContainer.
+type activateOneContainerStateMachine struct {
+	container StorageContainer // The associated storage contaier.
+
+	// cfg is the configuration for this activation, including those
+	// supplied to NewActivateContext (which apply to all activations),
+	// and those supplied to ActivateContext.ActivateContainer (which
+	// inherits the global configuration).
+	cfg activateConfig
+
+	stderr io.Writer // For writing error messages to.
+
+	next activateOneContainerStateMachineTask // The next state to run.
+
+	err            error                            // The first fatal error for this statemachine
+	status         ActivationStatus                 // Whether and how this container is activated.
+	keyslotRecords map[string]*keyslotAttemptRecord // Keyslot specific status.
```

Comment:
the comment probably should explain what the key is? the keyslot name?

Reply:
I can add a comment for this.